src/sage/schemes/elliptic_curves/cm.py

```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
-            #   lb(n) = n/(exp(euler_gamma)*log(log(n)) + 3/log(log(n)))
-            #   plot(lb, (n, 1, 10^4), color='red') + plot(lambda x: euler_phi(int(x)), 1, 10^4).show()
-            #
-            # So we consider f=1,2,..., until the first f with lb(f)*h_D > c*h_max.
-            # (Note that lb(f) is <= 0 for f=1,2, so nothing special is needed there.)
-            #
-            # TODO: Maybe we could do better using a bound for phi_D(f).
-            #
-            f = Integer(1)
-            chmax=hmax
-            if D==-3:
-                chmax*=3
+    # To avoid recomputing class numbers, we initialise T with the
+    # global hDf_dict, only including those (D,f) for which |D|*f**2
+    # <= B if B was provided:
+
+    T = hDf_dict.copy()
+
+    if not count:
+        for h in T:
+            T[h] = [Df for Df in T[h] if Df[0].abs()*Df[1]**2<=B]
+
+    # h_dict caches the class number h of all discriminants previously
+    # encountered; we will use the function OrderClassNumber() to
+    # quicky compute the class number of non-fundamental discriminants
+    # from the fundamental ones. Note that in the initialisation, the
+    # keys of h_dict include nonfundamental discriminants, but we only
+    # update it with fundamental ones.
+
+    h_dict = {}
+    for h, Dflist in hDf_dict.items():
+        for D0,f in Dflist:
+            h_dict[D0*f**2] = h
+
+    # We do not need to certaify the class number from :pari:`qfbclassno` for discriminants under 2*10^10
+    if B > 2*10**10:
+        proof = False
+
+    for D in xsrange(-3, -B-1, -1):
+        if not D.is_discriminant():
+            continue
+        D0 = D.squarefree_part()
+        if D0%4 !=1:
+            D0 *= 4
+        f = (D//D0).isqrt()
+
+        # Now D0 is the fundamental discriminant and f the conductor
+
+        if D in h_dict:
+            h = h_dict[D]
+        else:
+            if f == 1: # D itself is fundamental
+                h = D.class_number(proof)
+                h_dict[D] = h
             else:
-                if D==-4:
-                    chmax*=2
-            while lb(f)*h_D <= chmax:
-                if f == 1:
-                    h = h_D
-                else:
-                    h = (D*f*f).class_number(proof)
-                # If the class number of this order is within the range, then
-                # use it.  (NOTE: In some cases there is a simple relation between
-                # the class number for D and D*f^2, and this could be used to
-                # optimize this inner loop a little.)
-                if h <= hmax:
-                    z = (D, f)
-                    if h in T:
-                        T[h].append(z)
-                    else:
-                        T[h] = [z]
-                f += 1
+                h = OrderClassNumber(D0,h_dict[D0],f)
+
+        # If the class number of this order is within the range, then store (D0,f)
+        if h <= hmax:
+            z = (D0, f)
+            if h in T:
```

Comment:
you could use `T = defaultdict(set)` and then this would just be `T[h].add(z)`.

Reply:
This is a very neat idea -- and I did not know about defaultdicts, so thanks for that -- but it just required too much additional coding everywhere else, since I rely on the values of hDf_dict being sorted lists, not sets.

Of course I am willing to see a solution based on this idea!

Reply:
> This is a very neat idea -- and I did not know about defaultdicts, so thanks for that -- but it just required too much additional coding everywhere else, since I rely on the values of hDf_dict being sorted lists, not sets.
> 
> Of course I am willing to see a solution based on this idea!

OK, the last commit does this (11 insertions(+), 16 deletions(-) :))

Reply:
Great!  I use defaultdicts a lot, so I'm glad the suggestion ended up being useful.